src/Workspaces/Remote/ServiceHub/Shared/Extensions.cs

```diff
@@ -0,0 +1,108 @@
+                    RaiseCancellationIfInvokeFailed(task, mergedCancellation, cancellationToken);
+
+                    // wait for asset source to respond
+                    await stream.WaitForDirectConnectionAsync(mergedCancellation.Token).ConfigureAwait(false);
+
+                    // run user task with direct stream
+                    await funcWithDirectStreamAsync(stream, mergedCancellation.Token).ConfigureAwait(false);
+
+                    // wait task to finish
+                    await task.ConfigureAwait(false);
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                // if cancelled due to us, throw cancellation exception.
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // if canclled due to invocation is failed, rethrow merged cancellation
+                throw;
+            }
+        }
+
+        public static async Task<T> InvokeAsync<T>(
+            this JsonRpc rpc, string targetName, IEnumerable<object> arguments,
+            Func<Stream, CancellationToken, Task<T>> funcWithDirectStreamAsync, CancellationToken cancellationToken)
+        {
+            Task task = null;
+
+            try
+            {
+                using (var mergedCancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken))
+                using (var stream = new ServerDirectStream())
+                {
+                    // send request to asset source
+                    task = rpc.InvokeAsync(targetName, arguments.Concat(stream.Name).ToArray());
+
+                    // if invoke throws an exception, make sure we raise cancellation.
+                    RaiseCancellationIfInvokeFailed(task, mergedCancellation, cancellationToken);
+
+                    // wait for asset source to respond
+                    await stream.WaitForDirectConnectionAsync(mergedCancellation.Token).ConfigureAwait(false);
+
+                    // run user task with direct stream
+                    var result = await funcWithDirectStreamAsync(stream, mergedCancellation.Token).ConfigureAwait(false);
+
+                    // wait task to finish
+                    await task.ConfigureAwait(false);
+
+                    return result;
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                // if cancelled due to us, throw cancellation exception.
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // if canclled due to invocation is failed, rethrow merged cancellation
+                throw;
+            }
+        }
+
+        private static void RaiseCancellationIfInvokeFailed(Task task, CancellationTokenSource mergedCancellation, CancellationToken cancellationToken)
+        {
+            // if invoke throws an exception, make sure we raise cancellation
+            var dummy = task.ContinueWith(_ =>
```

Comment:
I can't wait on it since that will be dead lock. this gets data from stream while task is not fishined. otherside will not finish until it finish sending all data.